From 26e2c7e8e4b49b9b81147d44254216261390538a Mon Sep 17 00:00:00 2001
From: Bai Ping <b51503@freescale.com>
Date: Wed, 3 Jun 2015 22:36:27 +0800
Subject: [PATCH] MLK-11021-02 ARM: clk: imx6: update the perclk and uart clk

Use 24MHz osc clk as the perclk's parent clk to make sure the
perclk rate can't be change when we change the ipg clk.

For uart, change the clk source of uart_sel according the command line
argument. pass 'uart_from_osc' on command line to select the 24MHz osc
as the clk source of uart.

Signed-off-by: Bai Ping <b51503@freescale.com>
---
 arch/arm/mach-imx/clk-imx6ul.c | 11 ++++++++++-
 1 file changed, 10 insertions(+), 1 deletion(-)

diff --git a/arch/arm/mach-imx/clk-imx6ul.c b/arch/arm/mach-imx/clk-imx6ul.c
index c7213f8..1360c2e 100644
--- a/arch/arm/mach-imx/clk-imx6ul.c
+++ b/arch/arm/mach-imx/clk-imx6ul.c
@@ -155,7 +155,7 @@ static void __init imx6ul_clocks_init(struct device_node *ccm_node)
 	imx_clk_set_parent(clks[IMX6UL_PLL6_BYPASS], clks[IMX6UL_CLK_PLL6]);
 	imx_clk_set_parent(clks[IMX6UL_PLL7_BYPASS], clks[IMX6UL_CLK_PLL7]);
 
-	clks[IMX6UL_CLK_PLL1_SYS]	= imx_clk_gate("pll1_sys", 	"pll1_bypass", base + 0x00, 13);
+	clks[IMX6UL_CLK_PLL1_SYS]	= imx_clk_fixed_factor("pll1_sys",	"pll1_bypass", 1, 1);
 	clks[IMX6UL_CLK_PLL2_BUS]	= imx_clk_gate("pll2_bus", 	"pll2_bypass", base + 0x30, 13);
 	clks[IMX6UL_CLK_PLL3_USB_OTG]	= imx_clk_gate("pll3_usb_otg", 	"pll3_bypass", base + 0x10, 13);
 	clks[IMX6UL_CLK_PLL4_AUDIO]	= imx_clk_gate("pll4_audio", 	"pll4_bypass", base + 0x70, 13);
@@ -407,6 +407,15 @@ static void __init imx6ul_clocks_init(struct device_node *ccm_node)
 	clk_register_clkdev(clks[IMX6UL_CLK_GPT1_SERIAL], "per", "imx-gpt.0");
 	clk_register_clkdev(clks[IMX6UL_CLK_GPT_3M], "gpt_3m", "imx-gpt.0");
 
+	/* set perclk to from OSC */
+	imx_clk_set_parent(clks[IMX6UL_CLK_PERCLK_SEL], clks[IMX6UL_CLK_OSC]);
+
+	/* Set the UART parent iif needed */
+	if (uart_from_osc)
+		imx_clk_set_parent(clks[IMX6UL_CLK_UART_SEL], clks[IMX6UL_CLK_OSC]);
+	else
+		imx_clk_set_parent(clks[IMX6UL_CLK_UART_SEL], clks[IMX6UL_CLK_PLL3_80M]);
+
 	imx_clk_set_rate(clks[IMX6UL_CLK_ENET_REF], 50000000);
 	imx_clk_set_rate(clks[IMX6UL_CLK_ENET2_REF], 50000000);
 
-- 
1.8.0

